### DISCUSSION TITLE
Decryption problem

### DISCUSSION THREAD

**Discussion by spelunkerofbasil** [opendreamnet/dreamlink#31] [Q&A] [answered]
I'm active in the Art NFT community and am looking for a way to easily and reliably upload an encrypted Zip file of an image that I want to be accessible ONLY by the purchaser of an NFT.  I'm doing this as a way to make an image file TRULY rare, so that the purchaser is getting something entirely unique and unavailable elsewhere, and not just ownership bragging rights.  I thought that putting the image on IPFS and pinning it might be the best way to ensure that the file would be available whether it sells in a day or in a year.

Your project looks promising for this use, but I'm having a problem with decryption after download.  I have been testing it with a small image.  The first time I tried, it downloaded without issue, but I couldn't decrypt it.  The second time, it worked.  Then I had a friend download it (I use a Mac and he uses a PC.) and he also could not decrypt the file.  So I downloaded it again and it decrypted just fine.
 
The test file is here:
 
https://www.dreamlink.cloud/explorer?cid=Qmcr6m9hsEATWrfkNVERkQVxbG26PHCLTYqttY4KWd6pU3&filename=Adrift.zip 

The Zip encryption key is “alec2” (no quotes). 

So my questions are: (sorry for being stupid, but I'm not a coder) 

- Do you think the zip file could somehow become damaged in the distribution process?  I doubt that, because it has worked for me, but there are a lot of variables involved. 
- Do you think there might be some other issue at work creating the problem?  
- Do you have any suggestions for me?

Whatever solution I end up using will need to be bullet-proof, since the buyer may have even less computer skill than do I, and would likely be REALLY pissed if he couldn't get what he paid for!

Thanks!

**Comment by kolessios** [comment, marked as answer]
Hello, thanks for your interest in the project and an apology for the late reply, I had to rest several days for medical reasons.

Your problem seems to be in the way MacOS compresses and encrypts the .zip file, since when downloading it on my computer the native Windows decompressor fails but when using [7zip](https://www.7-zip.org/) the file decompresses correctly.

This makes me think that MacOS encrypts the file using `AES` which is not compatible with Windows, **it is recommended to use `ZipCrypto`**.

Here are 2 copies of your image, compressed with 7zip and encrypted with the same password but one with `AES` and one with `ZipCrypto`. The one encrypted with `AES` gives me the same error as your original file (in Windows).

https://www.dreamlink.cloud/explorer?cid=QmWvgbYTySRQcigc8KACeEZWNBscWpWBBusSjAAaCVNpSm&filename=Adrift%20Examples
_or_
https://bafybeid7svqfuiqsswn6n5e42vxqyuzbm4sgr4ef347n3tdrekrhssa6di.ipfs.dweb.link/

So the solution is to encrypt your files using `ZipCrypto`, this way any user on any operating system will be able to decompress the zip without additional programs.

I hope this can help you to continue with your goals. 🙂

As a side note, in IPFS it is very rare for a file to become corrupt, nodes always make sure to download the entire file and validate that it is not corrupt before distributing it.

**Reply by spelunkerofbasil** [reply]
Thank you so much.  I continued testing yesterday and arrived at a similar conclusion, at least in deciding that it was likely a problem with my encryption on the Mac--but not with the clarity of your analysis. I had initially used a trial version of WinZip to do the encryption, thinking that would provide best compatibility. It actually might; I may just have not configured it correctly. 

After following a couple of other roads to dead ends, I tried to do it via the command line, but getting the paths right was daunting for me, and advice on the www about it was all over the place, and there were even some heated discussions between coders.  So that was no help.

Eventually, I landed on Keka, which was inexpensive and had good documentation to the point that I understood the hazards of using AES, etc., and was finally successful in creating an encrypted zip that both I and my testing partner with a PC could successfully open and decrypt without a problem.

Your analysis nailed it.  I am most appreciative.
